opennms-dao/src/main/resources/META-INF/opennms/applicationContext-dao.xml

```diff
@@ -323,6 +323,24 @@
   
   <onmsgi:service interface="org.opennms.netmgt.dao.api.NodeDao" ref="nodeDao" />
 
+  <bean id="topologyEntityDao" class="org.opennms.netmgt.dao.hibernate.TopologyEntityDaoHibernate">
+    <property name="sessionFactory" ref="sessionFactory" />
+  </bean>
+
+  <onmsgi:service interface="org.opennms.netmgt.dao.api.TopologyEntityDao" ref="topologyEntityDao" />
+
+  <bean id="topologyEntityCache" class="org.opennms.netmgt.dao.hibernate.TopologyEntityCacheImpl">
+    <property name="topologyEntityDao" ref="topologyEntityDao" />
+  </bean>
+
+  <onmsgi:service interface="org.opennms.netmgt.dao.api.TopologyEntityCache" ref="topologyEntityCache" />
+
+  <task:scheduled-tasks scheduler="topologyRefresherScheduler">
+    <task:scheduled ref="topologyEntityCache" method="refresh" fixed-delay="300000"/>
```

Comment:
This now means that we'll be loading *all* of the nodes and all of the CDP links from the database every 5 minutes, even if no one is using the topology app.

Perhaps we should change the strategy to using a cache that triggers a load/refresh when the data is accessed instead.

Reply:
We could even have a "Loading UI Screen" in this case.
So the first time anyuser navigates to the topology map, we show a "Loading or Initializing screen". If the cache is initialized (either via event or cache.isInitiazlized()) we hide the loading screen.

There is a `TopologyService` which may be used for requesting a refresh of the underlying cache(s).

Reply:
The task was to make the initial loading of the topology map faster. We did do this mainly by 2 means:

1. load lightweight entities with no references and only the necessary attributes instead of the full entity (like OnmsNode) with lazy loaded references
2. preload the lightweight entities and caching them. The cache is a global one, not bound to a user.

Basically you are saying we shouldn't do the preloading in 2. anymore but switch to a load on demand strategy. Therefor we will loose that time benefit. On my computer this is for 40,000 Nodes and 80,000 CdpLinks: ~1-2,5 seconds. This delay we would add each time the cache is refreshed - not only for the initial loading.

We have in general two ways to fill a cache:
1. synchronously on demand / lazy loading -> slower but no unnecessary loading.
1. Asynchronous / preloading -> fast at access time but potentially unnecessary loading.

In order to determine what is the more appropriate strategy for our use case, I think we should look at the costs and the benefits and put that into the context of a typical OpenNMS usage: how many users do work in parallel, how often are the topology maps accessed, is CPU and RAM a concern, is sparse usage of resources more important than usability / speed?

A very rough indication of how much data we hold in memory: I serialized the cache (40,000 Nodes and 80,000 CdpLinks) which amounted to ~12MB. I don't know what impact it has on the network traffic and the database itself.

=> we are talking about a tradeoff between 1-2 seconds of delay for every cache access that needs a refresh vs. 12 MB RAM + database impact.

@j-white, @mvrueden what is your take on these numbers? Shall we do loading on demand or preloading (as implemented now)?

Reply:
> @j-white, @mvrueden what is your take on these numbers? Shall we do loading on demand or preloading (as implemented now)?

At the moment I would opt for filling the cache on demand by default.
That way we don't add unnecessary load to the system if the topology is not used at all.

Besides that we could add a system property which enables a periodical reload of the data.
And also make the refresh-interval configurable. Important here is, that the refresh is only triggered n-seconds AFTER the first refresh was completed. Not every n seconds.

Reply:
While preloading does give us the best performance, it comes at the cost of loading unnecessarily for many of the installs that do not use the topology app.

I agree with the approach that @mvrueden proposes above.

Reply:
ok, I changed the code accordingly.